[SPARK-1989] [CORE] Exit executors faster if they get into a cycle of heavy GC

devaraj-kavali opened the pull request:
## What changes were proposed in this pull request?

With the patch changes, executors get started with -XX:GCTimeLimit and/or -XX:GCHeapFreeLimit options if they are not configured as part of the configuration  "spark.executor.extraJavaOptions".
## How was this patch tested?

I have verified this by checking the executor process command when I ran different spark applications. I verified the same in different deployement modes(Standalone, YARN, Mesos).

Commit:
[SPARK-1989] [CORE] Exit executors faster if they get into a cycle of
heavy GC

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented:
You can already specify this option, like a bunch of other potentially relevant JVM flags, via the JVM flags option. I don't think we need to plumb through a special value for it and actually force people to not use the standard flags.

srowen commented:
(PS I think it would be useful to perhaps set these JVM flags, if not already set, to some more conservative value. I think that's the idea in the JIRA.)

devaraj-kavali commented:
Thanks @srowen for checking this immediately, I will make the changes as per your explanation.

Commit:
Adding options -XX:GCTimeLimit and -XX:GCHeapFreeLimit to the executor
command when they are not configured as part of
"spark.executor.extraJavaOptions".

devaraj-kavali commented:
@srowen I have made the changes, Please have a look into this. Thanks

srowen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/SparkDeploySchedulerBackend.scala:
```diff
@@ -66,12 +66,20 @@ private[spark] class SparkDeploySchedulerBackend(
+    // Add GC Limit options if they are not present
+    val extraJavaOptsAsStr = extraJavaOpts.mkString(" ")
+    if (!extraJavaOptsAsStr.contains("-XX:GCTimeLimit")) {
+      extraJavaOpts :+= Utils.getGCTimeLimitOption
```
This should all probably be done in one place, and farther upstream , in the ... Launcher? @vanzin would that cover all the paths to launch an executor? the idea here is to add a GC time limit arg as a default if not specified, for all executors.

vanzin replied:
The launcher code is not really used for creating the command line for executors. It contains code that might be useful there, though.

For this particular case, I'd take a look at `CommandBuilderUtils.addPermGenSizeOpt`. That's very similar to what is needed here. At the very least, I'd create a helper method instead of repeating this code in 3 different places. 

You also need to double check that the option won't cause issues with non-Oracle JVMs.

Finally, instead of creating a string, you could instead use `extraJavaOpts.exists(_.startsWith("-XX:GCTimeLimit=")`.

devaraj-kavali replied:
Thanks @vanzin for your comments, I will update with the comments fix and also will verify with non-Oracle JVMs.

tgravescs commented:
why are we adding this option at all here?  users can specify it or any other options in the java options.  Every application could be different and want to set this differently.  Some apps may be I want to finish no matter how long it takes vs some are I want to fail very fast.

I would rather see more support added for admin/cluster level config file then hardcode java GC configs like this.

devaraj-kavali commented:
Thanks @tgravescs for the comment, users can still specify this gc params as part of the java opts. If the user doesn't specify these gc params then only we are adding with default values for executors instead of the relying on JVM default values.

tgravescs commented:
I understand that it can be overridden but I'm not sure we should be in the business of setting the GC flags for people.  You set them to one value now, someone comes around and wants another default or wants to add another GC flag, you have to handle different java vendors and versions and it just becomes a maintenance headache.

Also as a user now if I wanted the defaults that came with my java version I have to explicitly override these (which I don't know we set since they aren't documented).  

Now I realize Spark is setting -XX:MaxPermSize but I would also argue there we shouldn't be.  As far as I know that is only around for legacy reasons because Spark was originally doing it.  @vanzin correct me if you know otherwise.

srowen commented:
It's debatable, yeah. This one was suggested by @mateiz a long time ago. Max perm size was set because Spark jobs would generally fail with the default JVM settings. (Non-issue in Java 8 anyway). This is sort of in the same category -- failing faster under non-trivial GC pressure rather than locking up or grinding out a result. 

The change vs default would be to fail if spending 90% rather than 98% of time in GC, or, fail if less than 5% of the heap is available rather than 2%. The former default seems way too high to me, personally. I'm not sure about the latter.

I suppose the argument has to be that, for Spark, the JVM default just doesn't make sense and needs a different default. I could believe ~90% makes sense at least for the first arg. I don't feel very strongly about it. Do you all?

vanzin commented:
re: perm gen, what Sean said. Spark just fails with the default value.

I don't feel strongly about adding this option or not. I have never seen anyone explicitly set it, so if it helps get stuck jobs move faster (or fail faster) I think it's a win.

tgravescs commented:
I understand the argument of we want the best user experience and I'm not against the settings themselves, I just think the benefit isn't worth the cost here. 

These are very specific advanced java options and properly maintaining and parsing them to me is not a necessary thing.  For instance when java 9,10,11 come out and the options no longer exist or change we have to go change code, if ibm java comes out with different config we have to change, if someone thinks 80% is better then 90% we have to change.  We already have enough PRs.

Let the user/admins configure it for their version of java and specific needs.  We are adding a bunch of code to parse these and set them to a default that someone thinks is better.  Many others might disagree. For instance with MapReduce we run it at 50% to fail fast.   Why not set spark to that?  if we want it to fail fast 50% is better then 90, right? Why don't we set the garbage collector as well?   To me this all comes down to configuring what is best for your specific application.  Since Spark can do so many different things - streaming, ML, graph processing, ETL, having one default isn't necessarily best for all. 

I think putting this in sets a bad precedence and just adds maintenance headache for not much benefit.  @vanzin mentions he has never seen anyone set this, so is it that big of a deal?  Where is the data that says 90% is better then 98% for the majority of Spark users.  Obviously if things just don't run like you mention with the max perm size, that makes it a much easier call and it makes sense to put it in, but I don't see that here.
Many of my customers don't set it and things are fine. I see other users set it because they explicitly want to fail very fast and its less then 90%.
  
I also think setting XX:GCHeapFreeLimit is more risky then setting GCTimeLimit. I personally have never seen anyone actually set this.  its defined as "The lower limit on the amount of space freed during a garbage collection in percent of the maximum heap (default is 2)" This to me is much more application specific then the GC time limit.

srowen commented:
I'm OK with not doing this, I think the contributor was just following up on an old idea from Matei.

tgravescs commented:
thanks for pointing that out, I made similar comments on the jira in case others don't agree with me we can discuss more.

srowen commented:
Thank you @devaraj-kavali for implementing this to show its impact, but I think you can close this PR now.
